### PR TITLE
general: support for merged records

### LINKED ISSUE #1609: Supporting merging records
In legacy when a record A is merged into a record B, A is tagged as deleted (`980__c` is set to `DELETED`) the recid of B is copied into `970__d` and when a user visit record A she is redirected to record B.

We need to implement the same behavior on master, and migrate existing merged information. I.e.:
- [x] We need to add to all schemas (when missing) a new field:

```
"new_record": {
            "$ref": "elements/json_reference.json",
            "description": "Master record that replaces this record",
            "title": "New record"
        },
```
- [x] We need to add rules for dojson that copy `970__d` into the proper URL in the above field. (e.g. http://localhost:5000/literature/1234)
- [x] We need to capture upon record insertion/update (i.e. with a receiver) when need to amend all the PIDs pointing to record A and make them redirect to record B.
- [x] BONUS POINT: schedule a task that look for all the records that were pointing to A and update them to point to B. Yay! (Actually see #484)

### PR DESCRIPTION
**Pull request by spirosdelviniotis**
* Closes #1609 
* Minor test refactoring.

Signed-off-by: Spiros Delviniotis <spyridon.delviniotis@cern.ch>

### CODE REVIEW

File: inspirehep/modules/migrator/tasks/records.py [hunk 1]
```diff
@@ -51,11 +51,13 @@
 from invenio_search.utils import schema_to_index
 
 from inspirehep.dojson.processors import overdo_marc_dict
+from inspirehep.dojson.utils import get_recid_from_ref
 from inspirehep.modules.pidstore.providers import InspireRecordIdProvider
 from inspirehep.modules.pidstore.minters import inspire_recid_minter
 from inspirehep.utils.dedupers import dedupe_list
 from inspirehep.utils.helpers import force_force_list
-from inspirehep.utils.record import get_value, soft_delete_pidstore_for_record
+from inspirehep.utils.record import get_value, soft_delete_pidstore_for_record, merge_pidstores_of_two_merged_records
```
Comment (jacquerie) [inline inspirehep/modules/migrator/tasks/records.py]: This should probably go on multiple lines:
```python
from inspirehep.utils.record import (
    get_value,
    merge_pidstores_of_two_merged_records,
    soft_delete_pidstore_for_record,
)
```

File: inspirehep/modules/records/jsonschemas/records/authors.json [hunk 2]
```diff
@@ -0,0 +1,274 @@
+{
```
Comment (jacquerie) [inline inspirehep/modules/records/jsonschemas/records/authors.json]: Here's one case in which an automatic rebase gets it wrong : )

If you are doing schema modifications, you have to submit a PR to https://github.com/inspirehep/inspire-schemas.

File: inspirehep/modules/records/receivers.py [hunk 3]
```diff
@@ -344,6 +344,31 @@ def check_if_record_is_going_to_be_deleted(sender, *args, **kwargs):
...
+        url = ref_obj.get('$schema', '')
+        if not url == '':
+            if 'hep' in url:
+                res = 'literature'
+            elif 'authors' in url:
+                res = 'authors'
+            else:
+                res = None
+        else:
+            res = None
```
Comment (jacquerie) [inline inspirehep/modules/records/receivers.py]: w00t!

I'm sure there is a better way to write this helper. It looks suspiciously close to #1612 and to what #1355 requests, so at least this needs a FIXME comment similar to those I mention there.

File: tests/integration/test_record.py [hunk 4]
```diff
@@ -98,6 +98,169 @@ def record_not_yet_deleted(app):
...
+                    r = RecordIndexer()
+                    r.index(record_merged)
+                    r.index(record_deleted)
+                    es.indices.refresh('records-hep')
+        db.session.commit()
+
+    yield
+
+    with app.app_context():
+        _delete_merged_records_from_everywhere('literature', 111, 222)
```
Comment (jacquerie) [inline tests/integration/test_record.py]: Refactoring opportunity: a meta-fixture that takes as argument two JSON records and generates a fixture like this one and the one above. Not urgent, but this repetition stinks a bit.

File: tests/integration/test_record.py [hunk 5]
```diff
@@ -115,6 +278,34 @@ def _delete_record_from_everywhere(collection, record_control_number):
     db.session.commit()
 
 
+def _delete_merged_records_from_everywhere(collection, record_control_number_1, record_control_number_2):
+    merged = PersistentIdentifier.query.all()[-2]
+    deleted = PersistentIdentifier.query.all()[-1]
```
Comment (jacquerie) [inline tests/integration/test_record.py]: Uh? How can this work?!

File: tests/integration/test_record.py [hunk 6]
```diff
@@ -133,3 +324,27 @@ def test_record_can_be_deleted(app, record_not_yet_deleted):
         db.session.commit()
 
         assert client.get('/api/literature/333').status_code == 410
+
+
+def test_insert_merged_deleted_record(app, records_already_merged_in_marcxml):
```
Comment (jacquerie) [inline tests/integration/test_record.py]: Nit: `test_merged_records_stay_merged`

File: tests/integration/test_record.py [hunk 7]
```diff
@@ -133,3 +324,27 @@ def test_record_can_be_deleted(app, record_not_yet_deleted):
...
+    with app.test_client() as client:
+        assert client.get('/api/literature/111').status_code == 200
+        assert client.get('/api/literature/222').status_code == 301
+
+
+def test_record_can_be_merged(app, records_not_merged_in_marcxml):
```
Comment (jacquerie) [inline tests/integration/test_record.py]: Nit: `test_records_can_be_merged`

File: inspirehep/utils/record.py [hunk 8]
```diff
@@ -143,6 +144,13 @@ def soft_delete_pidstore_for_record(record_id):
...
-        deleted = pid.delete()
+        pid.delete()
+
+
+def merge_pidstores_of_two_merged_records(record_merged_uuid, record_deleted_uuid):
+    """Merge all pidstores of deleted record to the merged record."""
+    pids_deleted = PersistentIdentifier.query.filter(PersistentIdentifier.object_uuid == record_deleted_uuid).all()
+    pids_merged = PersistentIdentifier.query.filter(PersistentIdentifier.object_uuid == record_merged_uuid).one()
 
-    return deleted
```
Comment (jacquerie) [inline inspirehep/utils/record.py]: BTW why this change? It should at least be in its own commit, but I wonder if here you're changing behaviour.

File: inspirehep/modules/records/jsonschemas/records/authors.json (hunk 2 above)
Comment (spirosdelviniotis) [inline inspirehep/modules/records/jsonschemas/records/authors.json]: ok! I have already done that! :smile: :

File: tests/integration/test_record.py (hunk 6 above)
Comment (spirosdelviniotis) [inline tests/integration/test_record.py]: :+1:

File: tests/integration/test_record.py (hunk 7 above)
Comment (spirosdelviniotis) [inline tests/integration/test_record.py]: :+1:

File: tests/integration/test_record.py (hunk 5 above)
Comment (spirosdelviniotis) [inline tests/integration/test_record.py]: Done!

File: inspirehep/utils/record.py (hunk 8 above)
Comment (spirosdelviniotis) [inline inspirehep/utils/record.py]: I've done it in a separate commit. It was useless because the returned value is not used somewhere. :smile:

**Comment by kaplun** [comment]
@spirosdelviniotis, please rebase and squash the 3 commits into a single one.
